content/en/blog/2024/otel-generative-ai/index.md

```diff
@@ -0,0 +1,89 @@
+
+**Spans: Tracing Model Interactions**  
+Spans track each model interaction’s lifecycle, covering input parameters (e.g., temperature, top_p) and response details like token count or errors. They provide visibility into each request, aiding in identifying bottlenecks and analyzing the impact of settings on model output.
+
+**Metrics: Monitoring Usage and Performance**  
+Metrics aggregate high-level indicators like request volume, latency, and resource use, essential for managing costs and performance. This data is particularly critical for API-dependent AI applications with rate limits and cost considerations.
+
+**Events: Capturing Detailed Interactions**  
+Events log detailed moments during model execution, such as user prompts and model responses, providing a granular view of model interactions. These insights are invaluable for debugging and optimizing AI applications where unexpected behaviors may arise.
+
+Together, these signals provide a comprehensive monitoring framework, enabling better cost management, performance tuning, and request tracing.
+
+**Extending Observability with Vendor-Specific Attributes**
+
+The Semantic Conventions also define vendor-specific attributes for platforms like OpenAI and Azure Inference API, ensuring telemetry captures both general and provider-specific details. This added flexibility supports multi-platform monitoring and in-depth insights. 
+
+## Building the Python Instrumentation Library for OpenAI
+
+This Python-based library for OpenTelemetry captures key telemetry signals for OpenAI models, providing developers with an out-of-the-box observability solution tailored to AI workloads. The library, [hosted within the OpenTelemetry Contrib project](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2), automatically collects telemetry from OpenAI model interactions, including request and response metadata, token usage, and operational metrics.
+
+As generative AI applications grow, additional instrumentation libraries for other languages will follow, extending OpenTelemetry support across more tools and environments. The current library’s focus on OpenAI highlights its popularity and demand within AI development, making it a valuable initial implementation.
+
+### Example Usage
+
+Here’s an example of using the OpenTelemetry Python library to monitor a generative AI application with the OpenAI client.
+Make sure you first install the library:
+
+```bash
+pip install opentelemetry-instrumentation-openai-v2
+```
+
+Then include the following code in your python application.
+
+```python
+from openai import OpenAI
+from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor
+
+OpenAIInstrumentor().instrument()
+
```

Comment:
we probably need to add the code that configure otel SDK (or add a comment pointing to docs on how to set it up).

The code snippet:

```python
# NOTE: OpenTelemetry Python Logs and Events APIs are in beta 
from opentelemetry import trace, _logs, _events
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._events import EventLoggerProvider

from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter

trace.set_tracer_provider(TracerProvider())
trace.get_tracer_provider().add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter())
)

_logs.set_logger_provider(LoggerProvider())
_logs.get_logger_provider().add_log_record_processor(BatchLogRecordProcessor(OTLPLogExporter()))
_events.set_event_logger_provider(EventLoggerProvider())
```

an existing example that does something similar - https://github.com/open-telemetry/opentelemetry-python/blob/stable/docs/examples/logs/example.py

/cc @lzchen @xrmx in case they have some suggestions

Reply:
That's a lot of extra code. Should I link to that code example or is there a better doc?

Reply:
let me add it to https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2988 and, once merged, let's just link it.